torch/_dynamo/convert_frame.py

```diff
@@ -508,6 +518,24 @@ def __call__(
         if is_generator(code):
             unimplemented("generator")
 
+        f_locals = frame.f_locals
+        if (
+            frame.f_code.co_name in ("__enter__", "__exit__")
+            and len(f_locals) == 1
+            and any(
+                type(a) is contextlib._GeneratorContextManager
+                for a in f_locals.values()
+            )
+        ):
+            ctx = next(
+                a
+                for a in f_locals.values()
+                if type(a) is contextlib._GeneratorContextManager
+            )
+            assert ctx is not None
+            skip_code(frame.f_code)
+            skip_code(ctx.gen.gi_frame.f_code)
```

Comment:
The skip code thing is not exactly what we want, but if the alternative (where we only enable the Generator logic if we're tracing through a context manager) doesn't work out then we can use it. The problem is that once the code has been skipped, it is skipped forever. This function sets some state on the code object:

https://www.internalfb.com/code/fbsource/[445fd165a7b25a5f1cb403aa45be653421be1ef1]/fbcode/caffe2/torch/csrc/dynamo/eval_frame.c?lines=828 

Concretely, the thing I'm worried about is if someone calls a function from the `__enter__` that is also used elsewhere, then the function used elsewhere will be skipped

Reply:
Actually, I misunderstood. We only need to skip two frames here (the enter and the frame of the contextlib.contextmanager-decorated function). It's ok that these are permanently skipped.

Reply:
We don't actually need to skip any code. There's a check above which raises unsupported if `code` is a generator. I had previously removed this check, but re-introduced it a few commits ago.
https://github.com/pytorch/pytorch/blob/82597d07aa5d06fb52bb748739a2c712a8d84810/torch/_dynamo/convert_frame.py#L508-L510